applications/asset_tracker/src/env_sensors/bsec.c

```diff
@@ -0,0 +1,271 @@
+/*
+ * Copyright (c) 2019 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-BSD-5-Clause-Nordic
+ */
+
+#include "bsec_integration.h"
+#include <i2c.h>
+#include <string.h>
+#include <zephyr.h>
+#include <atomic.h>
+#include <spinlock.h>
+#include <misc/byteorder.h>
+#include "env_sensors.h"
+#include <settings/settings.h>
+
+struct device *i2c_master;
+
+s64_t s_timestamp;
+float s_iaq;
+u8_t s_iaq_accuracy;
+float s_temperature;
+float s_humidity;
+float s_pressure;
+
+struct env_sensor {
+	env_sensor_data_t sensor;
+	struct k_spinlock lock;
+};
+
+static struct env_sensor temp_sensor = {
+	.sensor =  {
+		.type = ENV_SENSOR_TEMPERATURE,
+		.value = 0,
+	},
+};
+
+static struct env_sensor humid_sensor = {
+	.sensor =  {
+		.type = ENV_SENSOR_HUMIDITY,
+		.value = 0,
+	},
+};
+
+static struct env_sensor pressure_sensor = {
+	.sensor =  {
+		.type = ENV_SENSOR_AIR_PRESSURE,
+		.value = 0,
+	},
+};
+
+static struct env_sensor air_quality_sensor = {
+	.sensor =  {
+		.type = ENV_SENSOR_AIR_QUALITY,
+		.value = 0,
+	},
+};
+
+
+/* size of stack area used by bsec thread */
+#define STACKSIZE 4096
+
+/* scheduling priority used by bsec thread */
+#define PRIORITY CONFIG_SYSTEM_WORKQUEUE_PRIORITY
+
+static K_THREAD_STACK_DEFINE(thread_stack, STACKSIZE);
+static struct k_thread thread;
+
+u8_t s_state_buffer[BSEC_MAX_STATE_BLOB_SIZE];
+s32_t s_state_buffer_len;
+
+
+static int settings_set(const char *key, size_t len_rd,
+			settings_read_cb read_cb, void *cb_arg)
+{
+
+	if (!strcmp(key, "state")) {
+		s_state_buffer_len = len_rd;
+		if (read_cb(cb_arg, s_state_buffer, len_rd) > 0) {
+			return 0;
+		}
+	}
+	s_state_buffer_len = 0;
+	return -1;
+}
+
+static int enable_settings(void)
+{
+	int err = 0;
+
+	settings_subsys_init();
+	struct settings_handler my_conf = {
+		.name = "bsec",
+		.h_set = settings_set
+	};
+
+	err = settings_register(&my_conf);
+	if (err) {
+		printk("Cannot register settings handler");
+		goto error;
+	}
+
+	/* This module loads settings for all application modules */
+	err = settings_load();
+	if (err) {
+		printk("Cannot load settings");
+		goto error;
+	}
+
+error:
+	return err;
+}
+
+static int8_t bus_write(u8_t dev_addr, u8_t reg_addr,
+			u8_t *reg_data_ptr, u16_t data_len)
+{
+	u8_t buf[data_len+1];
+
+	buf[0] = reg_addr;
+	memcpy(&buf[1], reg_data_ptr, data_len);
+	return i2c_write(i2c_master, buf, data_len+1, dev_addr);
+}
+
+static s8_t bus_read(u8_t dev_addr, u8_t reg_addr,
+			u8_t *reg_data_ptr, u16_t data_len)
+{
+	return i2c_write_read(i2c_master, dev_addr, &reg_addr,
+				1, reg_data_ptr, data_len);
+}
+
+static void sleep(u32_t t_ms)
+{
+	k_sleep(t_ms);
+}
```

Comment:
Why the wrapper?

Reply:
Indeed.

Reply:
fixed. now (void *)k_sleep is passed to the functions instead.

Reply:
Fun fact: Did cost us a day of debugging when the api changed https://github.com/nrfconnect/sdk-nrf/pull/2353